test/Identity.IntegrationTest/Endpoints/IdentityServerTests.cs

```diff
@@ -0,0 +1,298 @@
+        }
+
+        [Fact]
+        public async Task TokenEndpoint_GrantTypeRefreshToken_Success()
+        {
+            var deviceId = "5a7b19df-0c9d-46bf-a104-8034b5a17182";
+            var username = "test+tokenrefresh@email.com";
+
+            await _factory.RegisterAsync(new RegisterRequestModel
+            {
+                Email = username,
+                MasterPasswordHash = "master_password_hash",
+            });
+
+            var (_, refreshToken) = await _factory.TokenFromPasswordAsync(username, "master_password_hash", deviceId);
+
+            var context = await _factory.Server.PostAsync("/connect/token", new FormUrlEncodedContent(new Dictionary<string, string>
+            {
+                { "grant_type", "refresh_token" },
+                { "client_id", "web" },
+                { "refresh_token", refreshToken },
+            }));
+
+            await AssertDefaultTokenBodyAsync(context);
+        }
+
+        [Fact]
+        public async Task TokenEndpoint_GrantTypeClientCredentials_Success()
+        {
+            var username = "test+tokenclientcredentials@email.com";
+            var deviceId = "8f14a393-edfe-40ba-8c67-a856cb89c509";
+
+            await _factory.RegisterAsync(new RegisterRequestModel
+            {
+                Email = username,
+                MasterPasswordHash = "master_password_hash",
+            });
+
+            var database = _factory.GetDatabaseContext();
+            var user = await database.Users
+                .FirstAsync(u => u.Email == username);
+
+            var context = await _factory.Server.PostAsync("/connect/token", new FormUrlEncodedContent(new Dictionary<string, string>
+            {
+                { "grant_type", "client_credentials" },
+                { "client_id", $"user.{user.Id}" },
+                { "client_secret", user.ApiKey },
+                { "scope", "api" },
+                { "DeviceIdentifier", deviceId },
+                { "DeviceType", ((int)DeviceType.FirefoxBrowser).ToString() },
+                { "DeviceName", "firefox" },
+            }));
+
+            using var body = await AssertHelper.ResponseIsAsync<JsonDocument>(context);
+            var root = body.RootElement;
+
+            Assert.Equal(JsonValueKind.Object, root.ValueKind);
+            AssertHelper.AssertJsonProperty(root, "access_token", JsonValueKind.String);
+            var expiresIn = AssertHelper.AssertJsonProperty(root, "expires_in", JsonValueKind.Number).GetInt32();
+            Assert.Equal(3600, expiresIn);
+            var tokenType = AssertHelper.AssertJsonProperty(root, "token_type", JsonValueKind.String).GetString();
+            Assert.Equal("Bearer", tokenType);
+            var scope = AssertHelper.AssertJsonProperty(root, "scope", JsonValueKind.String).GetString();
+            Assert.Equal("api", scope);
+        }
+
+        private static async Task<JsonDocument> AssertDefaultTokenBodyAsync(HttpContext httpContext)
```

Comment:
This method is a bit cryptic, what is "default token" body?

Reply:
Depending on the different grant type given the endpoint may include "extra" properties but the ones included here are what I have seen from every response and I thought it best to extract that into a function. Do you think `AssertBaseTokenBodyAsync` is better or any other ideas. I'm not married to the name.